espnet/nets/pytorch_backend/e2e_tts_transformer.py

```diff
@@ -0,0 +1,443 @@
+from espnet.nets.pytorch_backend.tacotron2.decoder import Prenet as DecoderPrenet
+from espnet.nets.pytorch_backend.tacotron2.encoder import Encoder as EncoderPrenet
+from espnet.nets.pytorch_backend.transformer.attention import MultiHeadedAttention
+from espnet.nets.pytorch_backend.transformer.decoder import Decoder
+from espnet.nets.pytorch_backend.transformer.embedding import PositionalEncoding
+from espnet.nets.pytorch_backend.transformer.embedding import ScaledPositionalEncoding
+from espnet.nets.pytorch_backend.transformer.encoder import Encoder
+from espnet.nets.pytorch_backend.transformer.layer_norm import LayerNorm
+from espnet.nets.pytorch_backend.transformer.plot import PlotAttentionReport
+from espnet.nets.tts_interface import TTSInterface
+from espnet.utils.cli_utils import strtobool
+
+
+class TransformerTTSLoss(torch.nn.Module):
+    """Transformer TTS loss function
+
+    :param Namespace args: argments containing following attributes
+        (bool) use_masking: whether to mask padded part in loss calculation
+        (float) bce_pos_weight: weight of positive sample of stop token (only for use_masking=True)
+    """
+
+    def __init__(self, args):
+        super(TransformerTTSLoss, self).__init__()
+        self.use_masking = args.use_masking
+        self.bce_pos_weight = args.bce_pos_weight
+
+    def forward(self, outs, logits, ys, labels, olens):
+        """Transformer loss forward computation
+
+        :param torch.Tensor outs: outputs i.e. log-mels (B, Lmax, odim)
+        :param torch.Tensor logits: stop logits (B, Lmax)
+        :param torch.Tensor ys: batch of padded target features (B, Lmax, odim)
+        :param torch.Tensor labels: batch of the sequences of stop token labels (B, Lmax)
+        :param list olens: batch of the lengths of each target (B)
+        :return: l1 loss value
+        :rtype: torch.Tensor
+        :return: mean square error loss value
+        :rtype: torch.Tensor
+        :return: binary cross entropy loss value
+        :rtype: torch.Tensor
+        """
+        # prepare weight of positive samples in cross entorpy
+        if self.bce_pos_weight != 1.0:
+            weights = ys.new(*labels.size()).fill_(1)
+            weights.masked_fill_(labels.eq(1), self.bce_pos_weight)
+        else:
+            weights = None
+
+        # perform masking for padded values
+        if self.use_masking:
+            mask = make_non_pad_mask(olens).unsqueeze(-1).to(ys.device)
+            ys = ys.masked_select(mask)
+            outs = outs.masked_select(mask)
+            labels = labels.masked_select(mask[:, :, 0])
+            logits = logits.masked_select(mask[:, :, 0])
+            weights = weights.masked_select(mask[:, :, 0]) if weights is not None else None
+
+        # calculate loss
+        l1_loss = F.l1_loss(outs, ys)
+        mse_loss = F.mse_loss(outs, ys)
+        bce_loss = F.binary_cross_entropy_with_logits(logits, labels, weights)
```

Comment:
```python
bce_loss = F.binary_cross_entropy_with_logits(logits, labels, pos_weight=self.bce_pos_weight)
```

Reply:
Wow, I did not know this option.
That is nice.